### ISSUE TITLE
[RFC] Check QEMU capabilities

### ISSUE THREAD

**Issue by Jakob-Naucke** [kata-containers/kata-containers#1709]
Kata assumes certain properties about the QEMU installation being used, mostly related to a recent enough version (and, as a consequence, the support of certain devices).
This can be secured to a degree through things like Snap, distro packaging, or static builds, and of course `versions.yaml` suggests a version too, but it is absolutely possible to run with an older version of QEMU and be greeted by QEMU's hopefully meaningful error messages. (They aren't always 100% helpful, e.g. when trying virtio-fs on s390x, QEMU <= 5.1.0 will tell you that the device `vhost-user-fs-ccw` is unknown, and it may or may not be clear to a user that they should upgrade QEMU or switch to 9p in Kata's settings.)
I think that it would be better to ensure QEMU can do the things you think it can do. 

First, there are different ways of doing this with different upsides and downsides.
- Check the QEMU version. Probably pretty fast, not particularly precise about devices.
- Check QEMU's `help` output for devices/objects/... you assume are available. Probably a bit slower, more precise about devices, but does not give you an answer as definite as...
- Do a device probe -- run a minimal QEMU (no machine, CPU stopped), query devices through QMP, and stop it again. This is very precise, but the time required to do this is not negligible. libvirt does this, but they can of course cache the result of that to avoid doing this on every VM launch. AFAICT, this isn't naturally easy to do in Kata, because there is no daemon. 

Then, there are also different places to check something like this.
- At the time of building Kata. This is generally problematic, since you cannot know about the system you would actually run on. 
- Through `kata-check`. This seems quite attractive, but can we ensure that it is run before a user might run Kata?
- When starting a sandbox. This is the most surefire way, but may have an impact on startup time.

I think we should discuss whether we should check QEMU at all, and if yes, when we should do which check.

/cc @kata-containers/architecture-committee, @fidencio, @devimc, @jcvenegas

**Comment by devimc** [comment]
@Jakob-Naucke thanks for filing this issue, IMO kata containers should print a message that includes the following sentence: `Please run "kata-runtime check" to test if system can run Kata Containers` only if `startSandbox` (run hypervisor) fails and implement the following point in the `check` sub-command.

> Do a device probe -- run a minimal QEMU (no machine, CPU stopped), query devices through QMP, and stop it again. This is very precise, but the time required to do this is not negligible. libvirt does this, but they can of course cache the result of that to avoid doing this on every VM launch. AFAICT, this isn't naturally easy to do in Kata, because there is no daemon.

it's responsibility of the user to check whether the hypervisor (custom or not) is compatible with kata before running a container.

**Comment by Jakob-Naucke** [comment]
@devimc that sounds reasonable.
> kata containers should print a message that includes the following sentence:

Do you mean at build time?

**Comment by alicefr** [comment]
@Jakob-Naucke In my experience, I won' rely on the devices reported by the help, especially for s390x

**Comment by devimc** [comment]
@Jakob-Naucke 
> Do you mean at build time?

no, I mean at runtime, when the hypervisor is started and it fails

**Comment by alicefr** [comment]
One additional point, even if you check the version, it depends if the devices were enabled at compile time in qemu. The only way to do a proper check is at runtime query the devices from qemu

**Comment by Jakob-Naucke** [comment]
I am currently looking at building a prototype for this.
I wonder whether we should e.g. check for `vhost-user-blk` (e.g. because it's unsupported on s390x), which isn't a configuration thing, but is used by passing a device to Docker according to the [doc](https://github.com/kata-containers/kata-containers/blob/54bcd3c9469bed526f88ec94385650b6da35e890/docs/use-cases/using-SPDK-vhostuser-and-kata.md) that we have on it. (Docker? Yes, this guide is written for Docker, and I'm not sure whether it's even possible these days. It doesn't appear to be in the tests.)
OTOH, that's not really something that happens accidentally.
/cc @devimc

**Comment by alicefr** [comment]
@Jakob-Naucke it is a very old issue https://github.com/kata-containers/runtime/issues/659
